File: tests/chobit_sexpr_tests.ts
import {
    NotAtomError,
    NotConsError,
    ReadError,
    WriteError,
    ValueType,
    ChobitSexpr,
    Iter
} from "../src/chobit_sexpr.ts";

function chobitSexprTest1() {
    const data = new Uint8Array([1, 2, 3, 4, 5]);

    const sexpr = ChobitSexpr.genAtom(data);
    console.assert(sexpr.isAtom());
    console.assert(!sexpr.isCons());

    const atom = sexpr.atom();
    console.assert(data.toString() == atom.toString());
}

function chobitSexprTest2() {
    const data1 = new Uint8Array([1, 2, 3, 4, 5]);
    const data2 = new Uint8Array([1, 2, 3, 4, 5, 6, 7]);

    const car = ChobitSexpr.genAtom(data1);
    const cdr = ChobitSexpr.genAtom(data2);

    const sexpr = ChobitSexpr.genCons(car, cdr);
    console.assert(!sexpr.isAtom());
    console.assert(sexpr.isCons());
    const car2 = sexpr.car();
    const cdr2 = sexpr.cdr();

    if (car2 && cdr2) {
        const carAtom = car2.atom();
        const cdrAtom = cdr2.atom();
        if (carAtom && cdrAtom) {
            console.assert(carAtom.toString() == data1.toString());
            console.assert(cdrAtom.toString() == data2.toString());
        }
    }
}

function chobitSexprTest3() {
    const data1 = new Uint8Array([1, 2, 3, 4, 5]);
    const data2 = new Uint8Array([1, 2, 3, 4, 5, 6, 7]);
    const data3 = new Uint8Array([1, 2, 3, 4, 5, 6, 7, 8, 9]);

    const item1 = ChobitSexpr.genAtom(data1);
    const item2 = ChobitSexpr.genAtom(data2);
    const item3 = ChobitSexpr.genAtom(data3);
    const nil = ChobitSexpr.genAtom(new Uint8Array(0));

    const sexpr = ChobitSexpr.genCons(
        item1,
        ChobitSexpr.genCons(
            item2,
            ChobitSexpr.genCons(
                item3,
                nil
            )
        )
    );
    console.log(sexpr);

    let current: ChobitSexpr = sexpr;
    while (current) {
        try {
            const car = current.car();

            console.assert(!current.isAtom());
            console.assert(current.isCons());

            const data = car.atom();
            console.log(data);
        } catch (e) {
            console.assert(e instanceof NotConsError);
            break;
        }

        current = current.cdr();
    }

    console.log(current);
}

function chobitSexprTest4() {
    {
        const value = 0x11;
        const atom = ChobitSexpr.genI8(value);
        console.assert(atom.readI8() == value);
        atom.writeI8(value + 1);
        console.assert(atom.readI8() == (value + 1));
    }

    {
        const value = 0x11;
        const atom = ChobitSexpr.genU8(value);
        console.assert(atom.readU8() == value);
        atom.writeU8(value + 1);
        console.assert(atom.readU8() == (value + 1));
    }

    {
        const value = 0x1111;
        const atom = ChobitSexpr.genI16(value);
        console.assert(atom.readI16() == value);
        atom.writeI16(value + 1);
        console.assert(atom.readI16() == (value + 1));
    }

    {
        const value = 0x1111;
        const atom = ChobitSexpr.genU16(value);
        console.assert(atom.readU16() == value);
        atom.writeU16(value + 1);
        console.assert(atom.readU16() == (value + 1));
    }

    {
        const value = 0x11111111;
        const atom = ChobitSexpr.genI32(value);
        console.assert(atom.readI32() == value);
        atom.writeI32(value + 1);
        console.assert(atom.readI32() == (value + 1));
    }

    {
        const value = 0x11111111;
        const atom = ChobitSexpr.genU32(value);
        console.assert(atom.readU32() == value);
        atom.writeU32(value + 1);
        console.assert(atom.readU32() == (value + 1));
    }

    {
        const value = BigInt("0x1111111111111111");
        const atom = ChobitSexpr.genI64(value);
        console.assert(atom.readI64() == value);
        atom.writeI64(value + BigInt(1));
        console.assert(atom.readI64() == (value + BigInt(1)));
    }

    {
        const value = BigInt("0x1111111111111111");
        const atom = ChobitSexpr.genU64(value);
        console.assert(atom.readU64() == value);
        atom.writeU64(value + BigInt(1));
        console.assert(atom.readU64() == (value + BigInt(1)));
    }

    {
        const value = 0.1234;
        const tmp = new Float32Array([value]);
        const atom = ChobitSexpr.genF32(tmp[0]);
        console.assert(atom.readF32() == tmp[0]);

        const tmp2 = new Float32Array([1.0]);
        const tmp3 = new Float32Array([tmp[0] + tmp2[0]]);
        atom.writeF32(tmp3[0]);
        console.assert(atom.readF32() == tmp3[0]);
    }

    {
        const value = 0.1234;
        const atom = ChobitSexpr.genF64(value);
        console.assert(atom.readF64() == value);
        atom.writeF64(value + 1.0);
        console.assert(atom.readF64() == (value + 1.0));
    }

    {
        const value = "Hello World";
        const atom = ChobitSexpr.genString(value);
        console.assert(atom.readString() == value);
    }
}

function chobitSexprTest5() {
    const data1 = new Uint8Array([1, 2, 3, 4, 5]);
    const data2 = new Uint8Array([1, 2, 3, 4, 5, 6, 7]);
    const data3 = new Uint8Array([1, 2, 3, 4, 5, 6, 7, 8, 9]);

    const item1 = ChobitSexpr.genAtom(data1);
    const item2 = ChobitSexpr.genAtom(data2);
    const item3 = ChobitSexpr.genAtom(data3);
    const nil = ChobitSexpr.genAtom(new Uint8Array(0));

    const sexpr = ChobitSexpr.genCons(
        item1,
        ChobitSexpr.genCons(
            item2,
            ChobitSexpr.genCons(
                item3,
                nil
            )
        )
    );

    for (const item of sexpr.iter()) {
        console.log(item.atom());
    }
}

function chobitSexprTest6() {
    const data1 = new Uint8Array([1, 2, 3, 4, 5]);
    const data2 = new Uint8Array([1, 2, 3, 4, 5, 6, 7]);
    const data3 = new Uint8Array([1, 2, 3, 4, 5, 6, 7, 8, 9]);

    const item1 = ChobitSexpr.genAtom(data1);
    const item2 = ChobitSexpr.genAtom(data2);
    const item3 = ChobitSexpr.genAtom(data3);
    const nil = ChobitSexpr.genAtom(new Uint8Array([]));

    const sexpr = ChobitSexpr.genCons(
        item1,
        ChobitSexpr.genCons(
            item2,
            ChobitSexpr.genCons(
                item3,
                nil
            )
        )
    );

    const ret1 = sexpr.carCdr();
    const [car1, cdr1] = ret1;
    const atom1 = car1.atom();
    console.assert(atom1.length == data1.length);
    for (let i in atom1) {
        console.assert(atom1[i] == data1[i]);
    }

    const ret2 = cdr1.carCdr();
    const [car2, cdr2] = ret2;
    const atom2 = car2.atom();
    console.assert(atom2.length == data2.length);
    for (let i in atom2) {
        console.assert(atom2[i] == data2[i]);
    }

    const ret3 = cdr2.carCdr();
    const [car3, cdr3] = ret3;
    const atom3 = car3.atom();
    console.assert(atom3.length == data3.length);
    for (let i in atom3) {
        console.assert(atom3[i] == data3[i]);
    }

    const atom4 = cdr3.atom();
    console.assert(atom4.length == 0);
}

function chobitSexprTest7() {
    {
        const value = 11;
        const sexpr = ChobitSexpr.genI8(value);
        const atom = sexpr.readI8();
        console.assert(atom == value);
    }

    {
        const value = 12;
        const sexpr = ChobitSexpr.genU8(value);
        const atom = sexpr.readU8();
        console.assert(atom == value);
    }

    {
        const value = 13;
        const sexpr = ChobitSexpr.genI16(value);
        const atom = sexpr.readI16();
        console.assert(atom == value);
    }

    {
        const value = 14;
        const sexpr = ChobitSexpr.genU16(value);
        const atom = sexpr.readU16();
        console.assert(atom == value);
    }

    {
        const value = 15;
        const sexpr = ChobitSexpr.genI32(value);
        const atom = sexpr.readI32();
        console.assert(atom == value);
    }

    {
        const value = 16;
        const sexpr = ChobitSexpr.genU32(value);
        const atom = sexpr.readU32();
        console.assert(atom == value);
    }

    {
        const value = BigInt(17);
        const sexpr = ChobitSexpr.genI64(value);
        const atom = sexpr.readI64();
        console.assert(atom == value);
    }

    {
        const value = BigInt(18);
        const sexpr = ChobitSexpr.genU64(value);
        const atom = sexpr.readU64();
        console.assert(atom == value);
    }

    {
        const value = new Float32Array(1);
        value[0] = 19.1;
        const sexpr = ChobitSexpr.genF32(value[0]);
        const atom = sexpr.readF32();
        console.assert(atom == value[0]);
    }

    {
        const value = new Float64Array(1);
        value[0] = 20.2;
        const sexpr = ChobitSexpr.genF64(value[0]);
        const atom = sexpr.readF64();
        console.assert(atom == value[0]);
    }
}

function chobitSexprTest8() {
    const data = new Uint8Array([1, 2, 3, 4, 5, 6, 7, 8, 9]);
    const sexpr = ChobitSexpr.genAtom(data);

    try {
        const atom = sexpr.readI8();

        console.assert(false);
    } catch (e) {
        if (e instanceof ReadError) {
            console.assert(e.valueType == ValueType.I8);
        } else {
            console.assert(false);
        }
    }

    try {
        const atom = sexpr.readU8();

        console.assert(false);
    } catch (e) {
        if (e instanceof ReadError) {
            console.assert(e.valueType == ValueType.U8);
        } else {
            console.assert(false);
        }
    }

    try {
        const atom = sexpr.readI16();

        console.assert(false);
    } catch (e) {
        if (e instanceof ReadError) {
            console.assert(e.valueType == ValueType.I16);
        } else {
            console.assert(false);
        }
    }

    try {
        const atom = sexpr.readU16();

        console.assert(false);
    } catch (e) {
        if (e instanceof ReadError) {
            console.assert(e.valueType == ValueType.U16);
        } else {
            console.assert(false);
        }
    }

    try {
        const atom = sexpr.readI32();

        console.assert(false);
    } catch (e) {
        if (e instanceof ReadError) {
            console.assert(e.valueType == ValueType.I32);
        } else {
            console.assert(false);
        }
    }

    try {
        const atom = sexpr.readU32();

        console.assert(false);
    } catch (e) {
        if (e instanceof ReadError) {
            console.assert(e.valueType == ValueType.U32);
        } else {
            console.assert(false);
        }
    }

    try {
        const atom = sexpr.readI64();

        console.assert(false);
    } catch (e) {
        if (e instanceof ReadError) {
            console.assert(e.valueType == ValueType.I64);
        } else {
            console.assert(false);
        }
    }

    try {
        const atom = sexpr.readU64();

        console.assert(false);
    } catch (e) {
        if (e instanceof ReadError) {
            console.assert(e.valueType == ValueType.U64);
        } else {
            console.assert(false);
        }
    }

    try {
        const atom = sexpr.readF32();

        console.assert(false);
    } catch (e) {
        if (e instanceof ReadError) {
            console.assert(e.valueType == ValueType.F32);
        } else {
            console.assert(false);
        }
    }

    try {
        const atom = sexpr.readF64();

        console.assert(false);
    } catch (e) {
        if (e instanceof ReadError) {
            console.assert(e.valueType == ValueType.F64);
        } else {
            console.assert(false);
        }
    }
}

console.log("chobitSexprTest1 ===========================================")
chobitSexprTest1();

console.log("chobitSexprTest2 ===========================================")
chobitSexprTest2();

console.log("chobitSexprTest3 ===========================================")
chobitSexprTest3();

console.log("chobitSexprTest4 ===========================================")
chobitSexprTest4();

console.log("chobitSexprTest5 ===========================================")
chobitSexprTest5();

console.log("chobitSexprTest6 ===========================================")
chobitSexprTest6();

console.log("chobitSexprTest7 ===========================================")
chobitSexprTest7();

console.log("chobitSexprTest8 ===========================================")
chobitSexprTest8();
console.log("============================================================")

File: tests/chobit_sexpr_tests.rs
extern crate chobitlibs;
extern crate turbo_json_checker;

use std::prelude::rust_2021::*;
use turbo_json_checker as tjc;

use chobitlibs::chobit_sexpr::*;
use core::mem::size_of;

#[test]
fn display_error_test() {
    assert!(tjc::validate_str(
        &ChobitSexprError::HeaderError(
            SexprHeaderError::CouldNotConvertFromSlice
        ).to_string()
    ).is_ok());

    assert!(tjc::validate_str(
        &ChobitSexprError::CouldNotRead(ValueType::I32).to_string()
    ).is_ok());

    assert!(tjc::validate_str(
        &ChobitSexprError::CouldNotWrite(ValueType::U64).to_string()
    ).is_ok());

    assert!(tjc::validate_str(
        &(ChobitSexprError::NotSexpr).to_string()
    ).is_ok());

    assert!(tjc::validate_str(
        &(ChobitSexprError::NotAtom).to_string()
    ).is_ok());

    assert!(tjc::validate_str(
        &(ChobitSexprError::NotCons).to_string()
    ).is_ok());
}

fn gen_test_data() -> Vec<u8> {
    const SIZE: usize = 1024;

    let mut ret = Vec::<u8>::with_capacity(SIZE);

    for i in 0..SIZE {
        ret.push(i as u8);
    }

    ret
}

#[test]
fn chobit_sexpr_test_1() {
    let data = gen_test_data();

    let buf = ChobitSexprBuf::new().push_atom(data.as_slice());
    let header = buf.as_sexpr().header().unwrap();

    assert!(header.is_atom());
    assert_eq!(header.size(), data.len());
    assert_eq!(buf.as_sexpr().atom().unwrap(), data.as_slice());
}

#[test]
fn chobit_sexpr_test_2() {
    const INDEX_1: usize = 10;
    const INDEX_2: usize = 30;

    let data = gen_test_data();

    let buf = ChobitSexprBuf::new()
        .build_cons()
        .push_car(
            ChobitSexprBuf::new().push_atom(&data[..INDEX_1]).as_sexpr()
        ).push_cdr(
            ChobitSexprBuf::new().push_atom(&data[INDEX_1..INDEX_2]).as_sexpr()
        );

    let header = buf.as_sexpr().header().unwrap();
    assert!(header.is_cons());
    assert_eq!(header.size(), INDEX_1 + size_of::<u32>());

    let sexpr = buf.as_sexpr().car().unwrap();
    let header = sexpr.header().unwrap();

    assert!(header.is_atom());
    assert_eq!(header.size(), INDEX_1);
    assert_eq!(sexpr.as_bytes().len(), INDEX_1 + size_of::<u32>());
    assert_eq!(sexpr.atom().unwrap(), &data[..INDEX_1]);

    let sexpr = buf.as_sexpr().cdr().unwrap();
    let header = sexpr.header().unwrap();

    assert!(header.is_atom());
    assert_eq!(header.size(), INDEX_2 - INDEX_1);
    assert_eq!(sexpr.as_bytes().len(), INDEX_2 - INDEX_1 + size_of::<u32>());
    assert_eq!(sexpr.atom().unwrap(), &data[INDEX_1..INDEX_2]);
}

#[test]
fn chobit_sexpr_test_3() {
    const INDEX_1: usize = 10;
    const INDEX_2: usize = 30;

    let data = gen_test_data();

    let buf = ChobitSexprBuf::new()
        .build_list()
        .push_item(
            ChobitSexprBuf::new().push_atom(&data[..INDEX_1]).as_sexpr()
        ).push_item(
            ChobitSexprBuf::new().push_atom(&data[INDEX_1..INDEX_2]).as_sexpr()
        ).finish();

    let header = buf.as_sexpr().header().unwrap();
    assert!(header.is_cons());
    assert_eq!(header.size(), INDEX_1 + size_of::<u32>());

    let sexpr = buf.as_sexpr().car().unwrap();
    let header = sexpr.header().unwrap();
    assert!(header.is_atom());
    assert_eq!(header.size(), INDEX_1);
    assert_eq!(sexpr.as_bytes().len(), INDEX_1 + size_of::<u32>());
    assert_eq!(sexpr.atom().unwrap(), &data[..INDEX_1]);

    let sexpr = buf.as_sexpr().cdr().unwrap();
    let header = sexpr.header().unwrap();
    assert!(header.is_cons());
    assert_eq!(header.size(), (INDEX_2 - INDEX_1) + size_of::<u32>());

    let sexpr = buf.as_sexpr().cdr().unwrap().car().unwrap();
    let header = sexpr.header().unwrap();
    assert!(header.is_atom());
    assert_eq!(header.size(), (INDEX_2 - INDEX_1));
    assert_eq!(sexpr.as_bytes().len(), (INDEX_2 - INDEX_1) + size_of::<u32>());
    assert_eq!(sexpr.atom().unwrap(), &data[INDEX_1..INDEX_2]);

    let sexpr = buf.as_sexpr().cdr().unwrap().cdr().unwrap();
    let header = sexpr.header().unwrap();
    assert!(header.is_atom());
    assert_eq!(header.size(), 0);
    assert_eq!(sexpr.as_bytes().len(), size_of::<u32>());
    assert_eq!(sexpr.atom().unwrap(), &[]);
}

#[test]
fn chobit_sexpr_test_4() {
    const INDEX_1: usize = 10;
    const INDEX_2: usize = 30;
    const INDEX_3: usize = 100;

    let data = gen_test_data();

    let buf = ChobitSexprBuf::new()
        .build_list()
        .push_item(
            ChobitSexprBuf::new().push_atom(&data[..INDEX_1]).as_sexpr()
        ).push_item(
            ChobitSexprBuf::new().push_atom(&data[INDEX_1..INDEX_2]).as_sexpr()
        ).finish_with(
            ChobitSexprBuf::new().push_atom(&data[INDEX_2..INDEX_3]).as_sexpr()
        );

    let header = buf.as_sexpr().header().unwrap();
    assert!(header.is_cons());
    assert_eq!(header.size(), INDEX_1 + size_of::<u32>());

    let sexpr = buf.as_sexpr().car().unwrap();
    let header = sexpr.header().unwrap();
    assert!(header.is_atom());
    assert_eq!(header.size(), INDEX_1);
    assert_eq!(sexpr.as_bytes().len(), INDEX_1 + size_of::<u32>());
    assert_eq!(sexpr.atom().unwrap(), &data[..INDEX_1]);

    let sexpr = buf.as_sexpr().cdr().unwrap();
    let header = sexpr.header().unwrap();
    assert!(header.is_cons());
    assert_eq!(header.size(), (INDEX_2 - INDEX_1) + size_of::<u32>());

    let sexpr = buf.as_sexpr().cdr().unwrap().car().unwrap();
    let header = sexpr.header().unwrap();
    assert!(header.is_atom());
    assert_eq!(header.size(), (INDEX_2 - INDEX_1));
    assert_eq!(sexpr.as_bytes().len(), (INDEX_2 - INDEX_1) + size_of::<u32>());
    assert_eq!(sexpr.atom().unwrap(), &data[INDEX_1..INDEX_2]);

    let sexpr = buf.as_sexpr().cdr().unwrap().cdr().unwrap();
    let header = sexpr.header().unwrap();
    assert!(header.is_atom());
    assert_eq!(header.size(), (INDEX_3 - INDEX_2));
    assert_eq!(sexpr.as_bytes().len(), (INDEX_3 - INDEX_2) + size_of::<u32>());
    assert_eq!(sexpr.atom().unwrap(), &data[INDEX_2..INDEX_3]);
}

#[test]
fn chobit_sexpr_test_5() {
    const INDEX_1: usize = 10;
    const INDEX_2: usize = 30;
    const INDEX_3: usize = 100;

    let data = gen_test_data();

    let mut buf = ChobitSexprBuf::new()
        .build_list()
        .push_item(
            ChobitSexprBuf::new().push_atom(&data[..INDEX_1]).as_sexpr()
        ).push_item(
            ChobitSexprBuf::new().push_atom(&data[INDEX_1..INDEX_2]).as_sexpr()
        ).finish_with(
            ChobitSexprBuf::new().push_atom(&data[INDEX_2..INDEX_3]).as_sexpr()
        );

    let new_data: Vec<u8> =
        data[INDEX_1..INDEX_2].iter().rev().map(|x| *x).collect();

    buf.cdr_mut().unwrap()
        .car_mut().unwrap()
        .atom_mut().unwrap()
        .copy_from_slice(&new_data);

    let header = buf.header().unwrap();
    assert!(header.is_cons());
    assert_eq!(header.size(), INDEX_1 + size_of::<u32>());

    let sexpr = buf.car().unwrap();
    let header = sexpr.header().unwrap();
    assert!(header.is_atom());
    assert_eq!(header.size(), INDEX_1);
    assert_eq!(sexpr.as_bytes().len(), INDEX_1 + size_of::<u32>());
    assert_eq!(sexpr.atom().unwrap(), &data[..INDEX_1]);

    let sexpr = buf.cdr().unwrap();
    let header = sexpr.header().unwrap();
    assert!(header.is_cons());
    assert_eq!(header.size(), (INDEX_2 - INDEX_1) + size_of::<u32>());

    let sexpr = buf.cdr().unwrap().car().unwrap();
    let header = sexpr.header().unwrap();
    assert!(header.is_atom());
    assert_eq!(header.size(), (INDEX_2 - INDEX_1));
    assert_eq!(sexpr.as_bytes().len(), (INDEX_2 - INDEX_1) + size_of::<u32>());
    assert_ne!(sexpr.atom().unwrap(), &data[INDEX_1..INDEX_2]);
    assert_eq!(sexpr.atom().unwrap(), &new_data);

    let sexpr = buf.cdr().unwrap().cdr().unwrap();
    let header = sexpr.header().unwrap();
    assert!(header.is_atom());
    assert_eq!(header.size(), (INDEX_3 - INDEX_2));
    assert_eq!(sexpr.as_bytes().len(), (INDEX_3 - INDEX_2) + size_of::<u32>());
    assert_eq!(sexpr.atom().unwrap(), &data[INDEX_2..INDEX_3]);
}

#[test]
fn chobit_sexpr_test_6() {
    let data = gen_test_data();

    let buf = ChobitSexprBuf::new().push_atom(data.as_slice());
    let header = buf.as_sexpr().header().unwrap();

    assert!(header.is_atom());
    assert_eq!(header.size(), data.len());
    assert_eq!(buf.as_sexpr().atom().unwrap(), data.as_slice());

    let new_data: Vec<u8> = data.iter().rev().map(|x| *x).collect();
    let buf = buf.clear().push_atom(&new_data);

    assert!(header.is_atom());
    assert_eq!(header.size(), new_data.len());
    assert_eq!(buf.as_sexpr().atom().unwrap(), new_data.as_slice());
    assert_ne!(buf.as_sexpr().atom().unwrap(), data.as_slice());
}

#[test]
fn chobit_sexpr_error_test_1() {
    let data = gen_test_data();

    let buf = ChobitSexprBuf::new().push_atom(data.as_slice());
    assert_eq!(Err(ChobitSexprError::NotCons), buf.car());
    assert_eq!(Err(ChobitSexprError::NotCons), buf.cdr());

    let mut new_data = buf.as_bytes().to_vec();
    let len = new_data.len();
    new_data.truncate(len - 1);

    let sexpr = ChobitSexpr::new(&new_data);
    assert_eq!(sexpr.as_bytes(), new_data.as_slice());
    assert_eq!(Err(ChobitSexprError::NotCons), sexpr.car());
    assert_eq!(Err(ChobitSexprError::NotCons), sexpr.cdr());
    assert_eq!(Err(ChobitSexprError::NotSexpr), sexpr.atom());
}

#[test]
fn chobit_sexpr_error_test_2() {
    const INDEX_1: usize = 10;
    const INDEX_2: usize = 30;

    let data = gen_test_data();

    let buf = ChobitSexprBuf::new()
        .build_cons()
        .push_car(
            ChobitSexprBuf::new().push_atom(&data[..INDEX_1]).as_sexpr()
        ).push_cdr(
            ChobitSexprBuf::new().push_atom(&data[INDEX_1..INDEX_2]).as_sexpr()
        );

    let mut new_data = buf.as_bytes().to_vec();
    new_data.truncate(data[..INDEX_1].len() + (size_of::<u32>() * 2) - 1);

    let sexpr = ChobitSexpr::new(&new_data);
    assert_eq!(Err(ChobitSexprError::NotSexpr), sexpr.car());
    assert_eq!(Err(ChobitSexprError::NotSexpr), sexpr.cdr());
    assert_eq!(Err(ChobitSexprError::NotAtom), sexpr.atom());

    let header = sexpr.header().unwrap();
    assert!(header.is_cons());
    assert_eq!(header.size(), new_data.len() - size_of::<u32>() + 1);
}

#[test]
fn chobit_sexpr_convert_test() {
    const COUNT: usize = 120;

    macro_rules! convert_test_core {
        ($type:ty, $num:expr) => {{
            let val: $type = $num as $type;

            let buf = ChobitSexprBuf::from(val);
            let result = <$type>::try_from(buf.as_sexpr()).unwrap();
            assert_eq!(result, val);

            let buf: ChobitSexprBuf<Completed> = val.into();
            let result: $type = buf.as_sexpr().try_into().unwrap();
            assert_eq!(result, val);
        }}
    }

    for i in 0..COUNT {
        convert_test_core!(i8, i);
        convert_test_core!(u8, i);
        convert_test_core!(i16, i);
        convert_test_core!(u16, i);
        convert_test_core!(i32, i);
        convert_test_core!(u32, i);
        convert_test_core!(i64, i);
        convert_test_core!(u64, i);
        convert_test_core!(i128, i);
        convert_test_core!(u128, i);
        convert_test_core!(f32, i);
        convert_test_core!(f64, i);
    }

    let val = "hello world";

    let buf = ChobitSexprBuf::from(val);
    let result = <&str>::try_from(buf.as_sexpr()).unwrap();
    assert_eq!(result, val);

    let buf: ChobitSexprBuf<Completed> = val.into();
    let result: &str = buf.as_sexpr().try_into().unwrap();
    assert_eq!(result, val);
}

#[test]
fn chobit_read_write_test() {
    macro_rules! chobit_read_write_test_core {
        ($read_func:ident, $write_func:ident, $type:ty, $value:expr) => {{
            let value = $value as $type;
            let mut sexpr = ChobitSexprBuf::from(value);

            assert_eq!(sexpr.as_sexpr().$read_func().unwrap(), value);

            sexpr.as_mut_sexpr().$write_func(value + (1 as $type)).unwrap();
            assert_eq!(
                sexpr.as_sexpr().$read_func().unwrap(),
                value + (1 as $type)
            );
        }};
    }

    for i in 0..100usize {
        chobit_read_write_test_core!(read_i8, write_i8, i8, i);
        chobit_read_write_test_core!(read_u8, write_u8, u8, i);
        chobit_read_write_test_core!(read_i16, write_i16, i16, i);
        chobit_read_write_test_core!(read_u16, write_u16, u16, i);
        chobit_read_write_test_core!(read_i32, write_i32, i32, i);
        chobit_read_write_test_core!(read_u32, write_u32, u32, i);
        chobit_read_write_test_core!(read_i64, write_i64, i64, i);
        chobit_read_write_test_core!(read_u64, write_u64, u64, i);
        chobit_read_write_test_core!(read_i128, write_i128, i128, i);
        chobit_read_write_test_core!(read_u128, write_u128, u128, i);
        chobit_read_write_test_core!(read_f32, write_f32, f32, i);
        chobit_read_write_test_core!(read_f64, write_f64, f64, i);
    }
}

#[test]
fn chobit_sexpr_iter_test() {
    let value_1: i32 = 100;
    let value_2: i32 = 200;
    let value_3: i32 = 300;

    let base = ChobitSexprBuf::new().build_list().push_item(
        ChobitSexprBuf::from(value_1).as_sexpr()
    ).push_item(
        ChobitSexprBuf::from(value_2).as_sexpr()
    ).push_item(
        ChobitSexprBuf::from(value_3).as_sexpr()
    ).finish();

    let add_value: i32 = 10;

    let mut buffer = ChobitSexprBuf::new().build_list();

    let mut tmp = ChobitSexprBuf::from(0i32);

    for elm in base.as_sexpr() {
        let value: i32 = elm.try_into().unwrap();
        tmp.write_i32(value + add_value).unwrap();

        buffer = buffer.push_item(&tmp);
    }

    let buffer = buffer.finish();

    let check = ChobitSexprBuf::new().build_list().push_item(
        ChobitSexprBuf::from(value_1 + add_value).as_sexpr()
    ).push_item(
        ChobitSexprBuf::from(value_2 + add_value).as_sexpr()
    ).push_item(
        ChobitSexprBuf::from(value_3 + add_value).as_sexpr()
    ).finish();

    assert_eq!(buffer.as_bytes(), check.as_bytes());
}

#[test]
fn chobit_sexpr_test_7() {
    let value_1: i32 = 100;
    let value_2: i32 = 200;
    let value_3: i32 = 300;

    let sexpr = ChobitSexprBuf::new().build_list().push_item(
        ChobitSexprBuf::from(value_1).as_sexpr()
    ).push_item(
        ChobitSexprBuf::from(value_2).as_sexpr()
    ).push_item(
        ChobitSexprBuf::from(value_3).as_sexpr()
    ).finish();

    let (car, cdr) = sexpr.car_cdr().unwrap();
    let value: i32 = car.try_into().unwrap();
    assert_eq!(value, value_1);

    let (car, cdr) = cdr.car_cdr().unwrap();
    let value: i32 = car.try_into().unwrap();
    assert_eq!(value, value_2);

    let (car, cdr) = cdr.car_cdr().unwrap();
    let value: i32 = car.try_into().unwrap();
    assert_eq!(value, value_3);

    assert_eq!(cdr.atom().unwrap(), &[]);
}

#[test]
fn push_sexpr_test() {
    let value_1: i32 = 100;
    let value_2: i32 = 200;
    let value_3: i32 = 300;

    let sexpr_1 = ChobitSexprBuf::new().build_list().push_item(
        ChobitSexprBuf::from(value_1).as_sexpr()
    ).push_item(
        ChobitSexprBuf::from(value_2).as_sexpr()
    ).push_item(
        ChobitSexprBuf::from(value_3).as_sexpr()
    ).finish();

    let sexpr_2 = ChobitSexprBuf::new().push_sexpr(&sexpr_1);

    assert_eq!(sexpr_2.as_sexpr(), sexpr_1.as_sexpr());
}

#[test]
fn push_number_test() {
    macro_rules! push_number_test_core {
        ($type:ty, $value:expr, $func_name:ident) => {{
            let value_1: $type = $value as $type;

            let sexpr = ChobitSexprBuf::new().$func_name(value_1);

            let value_2 = <$type>::try_from(sexpr.as_sexpr()).unwrap();

            assert_eq!(value_2, value_1);
        }};
    }

    for value in 0..100usize {
        push_number_test_core!(i8, value, push_i8);
        push_number_test_core!(u8, value, push_u8);
        push_number_test_core!(i16, value, push_i16);
        push_number_test_core!(u16, value, push_u16);
        push_number_test_core!(i32, value, push_i32);
        push_number_test_core!(u32, value, push_u32);
        push_number_test_core!(i64, value, push_i64);
        push_number_test_core!(u64, value, push_u64);
        push_number_test_core!(i128, value, push_i128);
        push_number_test_core!(u128, value, push_u128);
        push_number_test_core!(f32, value, push_f32);
        push_number_test_core!(f64, value, push_f64);
    }
}

#[test]
fn sexpr_fmt_test() {
    let value_1: u8 = 0x11;
    let value_2: u16 = 0x1122;
    let value_3: u32 = 0x11223344;
    let value_4: u64 = 0x1122334455667788;

    let sexpr = ChobitSexprBuf::new().build_list().push_item(
        ChobitSexprBuf::from(value_1).as_sexpr()
    ).push_item(
        ChobitSexprBuf::from(value_2).as_sexpr()
    ).push_item(
        ChobitSexprBuf::from(value_3).as_sexpr()
    ).push_item(
        ChobitSexprBuf::from(value_4).as_sexpr()
    ).finish();

    println!("{}", sexpr.as_sexpr());
}

#[test]
fn nil_test() {
    assert_eq!(ChobitSexpr::nil().atom().unwrap(), &[]);
}
